# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('C:/Users/daria/Датасеты для проектов/02 Предобработка данных - исследование надёжности заёмщиков/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

<a id='gap'></a>

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data['children'].unique() # Проверим, какие значения находятся в столбце 'children'
# чтобы было понимание, какую разбивку по категориям нужно сделать

array([1, 0, 3, 2, 4, 5], dtype=int64)

Благодаря поиску уникальных значений в столбце 'children' было принято решение разделить количество заемщиков на следующие группы:
- нет детей
- 1-2 ребенка
- многодетные

Такая категоризация объясняется следующим образом:

Клиенту без детей не затрудняется взять на себя обязательство по кредиту.
Многодетным клиент становится в ситуации, если у него 3 и более детей - на основе принятой в России классификации. 
С точки зрения платежеспособности многодетным клиентам труднее выплачивать кредит, чем тем, у кого 1-2 ребенка.

In [23]:
def category_children(child): # Возвращает группу по семье по значению количества детей в семье (child)
    if child < 1:
        return 'нет детей'
    if 1 <= child <= 2:
        return '1-2 ребенка'
    if child >= 3:
        return 'многодетные'

In [24]:
# Создали методом apply() новый столбец 'category_children' в датафрейме
data['category_children'] = data['children'].apply(category_children) 

Определили категории по количеству детей у потенциальных заемщиков, данные подготовлены к проверке гипотезы. Создаем сводную таблицу, которая покажет:
- **all_debts**

Общее количество клиентов, у которых была когда-либо задолженность по возврату кредитов либо нет. В задании не указано, подразумевается ли задолженность перед определенным банком, поэтому была принята во внимание кредитная история клиентов, где 0 - "идеальная".
- **arrears**

Количество клиентов, кредитная история которых неидеальная - была задолженность по кредитам.
- **debts_paid**

Количество клиентов, кредитная история которых идеальная - всегда платили по кредитам вовремя.
- **debts_paid_ratio**

Отношение клиентов, которые платили по долгам вовремя, к общему количеству заемщиков.

In [25]:
# сводная таблица, содержащая информацию по количеству и сумме строк по столбцу 'debt'
data_pivot = data.pivot_table(index = ['category_children'], values = 'debt', \
aggfunc = {'count', 'sum'}).rename(columns={'count':'all_debts', 'sum':'arrears'})
# добавляем столбец 'debts_paid' - разница между "общим количеством долгов" и "невыплаченными долгами"
data_pivot['debts_paid'] = data_pivot['all_debts'] - data_pivot['arrears']
# добавляем столбец 'debts_paid_ratio' - отношение выплаченных кредитов
# к общему количеству данных в столбце 'debt'
data_pivot['debts_paid_ratio'] = (data_pivot['debts_paid']/data_pivot['all_debts'])\
.apply(lambda x: round(x * 100, 2))
# для того, чтобы данные по процентам "читались" удобнее, была подсмотрена в гугле лямбда-функция 
data_pivot

all_debts  arrears  debts_paid  debts_paid_ratio
category_children                                                  
1-2 ребенка             6860      638        6222             90.70
многодетные              380       31         349             91.84
нет детей              14091     1063       13028             92.46

In [26]:
def data_pivot(data, aggr_index, vls, f1, f2, f1_rename, f2_rename, f3_name, f4_name):
    data_pivot = data.pivot_table(
        index = [aggr_index], 
        values = vls,
        aggfunc = {f1, f2}).rename(columns={f1:f1_rename, f2:f2_rename})
    data_pivot[f3_name] = data_pivot[f1_rename] - data_pivot[f2_rename]
    data_pivot[f4_name] = (data_pivot[f3_name]/data_pivot[f1_rename]).apply(lambda x: round(x * 100, 2))
    
    return data_pivot

new_data = data_pivot(data, 'category_children', 'debt', 'count', 'sum', 'all_debts', 'arrears',\
                      'debts_paid', 'debts_paid_ratio')
new_data

all_debts  arrears  debts_paid  debts_paid_ratio
category_children                                                  
1-2 ребенка             6860      638        6222             90.70
многодетные              380       31         349             91.84
нет детей              14091     1063       13028             92.46

**Вывод:** 

Зависимость между количеством детей и возвратом кредита в срок есть, хоть и различия в показателях несущественны:

1. Большую группу составляют **клиенты без детей**, которые с бОльшей вероятностью выплатят кредит, потому что отношение уплаченных долгов ко всем самый высокий - 92.46%. Возможно, это связано с тем, что семья, состоящая из двух человек, или холостой/незамужний клиент умеет распоряжаться бюджетом в рамках своего дохода, случайных издержек возникает реже.

2. **Многодетные семьи**, процент выплаты кредитов которых 91.84%, вероятнее всего это та категория людей, которые с большой ответственностью относится не только к воспитанию детей, но и к выплате кредитов. Кроме того, судя по данным сводной таблицы, количество клиентов, у которых более 3х детей гораздо ниже других категорий, поэтому данных по многодетным недостаточно для корректных выводов по ним.

3. Процент выплаты кредитов у потенциальных заемщиков, у которых есть **один или двое детей**, самый низкий - 90.70%. Это может быть по разным причинам, например, с рождением ребенка резко увеличиваются расходы на различные нужды, не всегда получается "рассчитать" бюджет.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [27]:
data['family_status'].value_counts() # Категории уже были в исходном датафрейме, посмотрим, какие:

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

Создаем сводную таблицу, в которой аналогично с группой по семье отобразим данные по группе по семейному положению:

In [28]:
data_pivot2 = data.pivot_table(index = ['family_status'], values = 'debt', \
aggfunc = {'count', 'sum'}).rename(columns={'count':'all_debts', 'sum':'arrears'})

data_pivot2['debts_paid'] = data_pivot2['all_debts'] - data_pivot2['arrears']

data_pivot2['debts_paid_ratio'] = (data_pivot2['debts_paid']/data_pivot2['all_debts'])\
.apply(lambda x: round(x * 100, 2))
 
data_pivot2

all_debts  arrears  debts_paid  debts_paid_ratio
family_status                                                          
Не женат / не замужем       2796      273        2523             90.24
в разводе                   1189       84        1105             92.94
вдовец / вдова               951       63         888             93.38
гражданский брак            4134      385        3749             90.69
женат / замужем            12261      927       11334             92.44

In [29]:
new_data2 = data_pivot(data, 'family_status', 'debt', 'count', 'sum', 'all_debts', 'arrears',\
                      'debts_paid', 'debts_paid_ratio')
new_data2

all_debts  arrears  debts_paid  debts_paid_ratio
family_status                                                          
Не женат / не замужем       2796      273        2523             90.24
в разводе                   1189       84        1105             92.94
вдовец / вдова               951       63         888             93.38
гражданский брак            4134      385        3749             90.69
женат / замужем            12261      927       11334             92.44

**Вывод:** 

Зависимость между семейным положением и возвратом кредита в срок есть:

- **"Вдовец/вдова"** имеют самый высокий процент возврата кредита - 93.38%.
- Примерно на одном уровне **"в разводе"** (92.94%) и **"женат/замужем"** (92.44%).
- Примерно на одному уровне **"гражданский брак"** (90.69%) и **"Не женат / не замужем"** (90.24%).

Объяснить данную ситуацию можно следующим образом:

Люди в категории "Вдовец/вдова" скорее всего люди в пожилом возрасте, где уровень ответственности высокий. Человек в разводе опирается сам на себя, поэтому можно сделать вывод, что расходы на непредвиденные ситуации будут малы, что позволит выплачивать по кредиту вовремя. У женатых/замужних есть поддержка, что также будет способствовать возврату кредита в срок. Что касается людей в гражданском браке и не женатых/ не замужних - по ним наименьших процент по выплатам в срок, возможно, на это влияет их статус, подразумевающий, что люди пока не готовы нести ответственность за другого человека, в том числе и за свои решения.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [30]:
data['total_income_category'].value_counts() # Категории уже были выявлены в датафрейме:

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

Для того, чтобы было удобнее финализировать вывод, назовем категории от А до Е:

- **E - "ниже среднего"**:
доход до 30.000

- **D - "низкий средний класс"**:
доход от 30.001 до 50.000

- **C - "высокий средний класс"**:
доход от 50.001 до 200.000

- **B - "богатые"**:
доход от 200.001 до 1.000.000

- **A - "сверхбогатые"**:
доход от 1.000.001

Создаем сводную таблицу, в которой аналогично с группой по семье отобразим данные по группе по уровню дохода:

In [31]:
new_data3 = data_pivot(data, 'total_income_category', 'debt', 'count', 'sum', 'all_debts', 'arrears',\
                      'debts_paid', 'debts_paid_ratio')
new_data3

all_debts  arrears  debts_paid  debts_paid_ratio
total_income_category                                                  
A                             25        2          23             92.00
B                           5014      354        4660             92.94
C                          15921     1353       14568             91.50
D                            349       21         328             93.98
E                             22        2          20             90.91

**Вывод:**

Зависимость между доходом и возвратом кредита в срок есть:

- Катагория D **"низкий средний класс"** имеет самый высокий процент возврата кредитов - 93.98%.
- Следом за ними клиенты, попадающие в категорию B **"богатые"**, с показателем 92.94%.
- Категория А или **"сверхбогатые"** клиенты с зарплатой более 1 миллиона имеют процент по возврату кредита - 92%, примерно такой же процент у клиентов, входящих в группу **"высокий средний класс"** (категория С) - 91.50%.
- Самый низкий показатель у категории клиентов Е, входящих в **"ниже среднего"** (90.91%).

Объяснить данную ситуацию можно следующим образом:

Клиенты с зарплатой от 30.001 до 50.000 лучше всех выплачивают кредиты, вероятнее всего это связано с ограниченным доходом, который не позволяет тратить средства импульсивно, и все возможные дополнительные расходы тщательно обдумываются или не совершаются вовсе.

Люди с высокой зарплатой выплачивают кредиты в большинстве случаев вовремя, потому что могут распределять доход по всем своим нуждам. Но все же показатель возврата кредита в срок у **высшего среднего класса ниже** (категория С) ниже, чем у клиентов, входящих в группу В - **богатые**. Возможно, более низкие показатели возвратов кредитов в срок у категории С связано с тем, что клиенты вышли на новый уровень жизни из-за увеличения зарплаты, и пока еще расходуют больше обычного.

Исходя из результатов, можно сделать вывод, что категории А и Е некорректно принимать к рассмотрению, так как данных по ним очень мало. Но с другой стороны, клиенты с зарплатой более 1 млн рублей могут редко обращаться к банку за кредитом именно на личные нужды. Как и клиенты с зарплатой до 30.000 - они могут взять микрозайм, чтобы решить свою потребность быстрее, обходя ужесточенные требования банков.

Создали 5 интервалов, посмотрим, как происходит распределение в группах: в каждом интервале находится равное количество данных.

In [32]:
pd.qcut(data['total_income'], q=5).value_counts()

(20666.999, 98514.0]     4267
(98514.0, 132113.0]      4266
(132113.0, 161380.0]     4266
(161380.0, 214604.0]     4266
(214604.0, 2265604.0]    4266
Name: total_income, dtype: int64

Из полученного наблюдения выше можно сделать вывод, что распределение заработной платы по категориям будет отличаться от использованного метода ранее, но обозначим аналогично категории с А до Е. 

- **E - ниже среднего**:
доход от 20666.999 до 98514.0

- **D - низкий средний класс**:
доход от 98514.0 до 132113.0

- **C - высокий средний класс**:
доход от 132113.0 до 161380.0

- **B - богатые**:
доход от 161380.0 до 214604.0

- **A - сверхбогатые**:
доход от 214604.0 до 2265604.0

In [33]:
groups = ['E', 'D', 'C', 'B', 'A']
pd.qcut(data['total_income'], q=5).value_counts()

data['total_income_category_new'] = pd.qcut(data['total_income'],
                              q=[0, .2, .4, .6, .8, 1],
                              labels=groups)
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  \
0               покупка жилья                     B  операции с недвижимостью   
1     приобретение автомобиля                     C    операции с автомобилем   
2               покупка жилья                     C  операции с недвижимостью   
3  дополнительное образование                     B     получение образования   
4             сыграть свадьбу                     C        проведение свадьбы   

  category_children total_income_category_new  
0       1-2 ребенка                         A  
1       1-2 ребенка                         D  
2         нет детей                         C  
3       многодетные                         A  
4         нет детей                         C

Создадим сводную таблицу для анализа связи между уровнем дохода и возвратом кредита в срок

In [34]:
new_data33 = data_pivot(data, 'total_income_category_new', 'debt', 'count', 'sum', 'all_debts', 'arrears',\
                      'debts_paid', 'debts_paid_ratio')
new_data33

all_debts  arrears  debts_paid  debts_paid_ratio
total_income_category_new                                                  
E                               4267      344        3923             91.94
D                               4266      358        3908             91.61
C                               4266      373        3893             91.26
B                               4266      358        3908             91.61
A                               4266      299        3967             92.99

**Вывод:**

В данном случае показатели возврата кредитов в срок у категорий B-E примерно одинаковое - 91.6%, а клиенты, входящие в группу A - в конкретном случае получающие зарплату от 214604.0 до 2265604.0 имеют наибольший процент возврата кредита вовремя - 92.99%.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [35]:
data['purpose_category'].value_counts() # Мы уже выделили основные цели кредита.

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

Теперь с помощью сводной таблицы определим, какая существует зависимость между целями кредита и возвратом кредита в срок:

In [36]:
data_pivot4 = data.pivot_table(index = ['purpose_category'], values = 'debt', \
aggfunc = {'count', 'sum'}).rename(columns={'count':'all_debts', 'sum':'arrears'})

data_pivot4['debts_paid'] = data_pivot4['all_debts'] - data_pivot4['arrears']

data_pivot4['debts_paid_ratio'] = (data_pivot4['debts_paid']/data_pivot4['all_debts'])\
.apply(lambda x: round(x * 100, 2))
 
data_pivot4

all_debts  arrears  debts_paid  debts_paid_ratio
purpose_category                                                          
операции с автомобилем         4279      400        3879             90.65
операции с недвижимостью      10751      780        9971             92.74
получение образования          3988      369        3619             90.75
проведение свадьбы             2313      183        2130             92.09

In [37]:
new_data4 = data_pivot(data, 'purpose_category', 'debt', 'count', 'sum', 'all_debts', 'arrears',\
                      'debts_paid', 'debts_paid_ratio')
new_data4

all_debts  arrears  debts_paid  debts_paid_ratio
purpose_category                                                          
операции с автомобилем         4279      400        3879             90.65
операции с недвижимостью      10751      780        9971             92.74
получение образования          3988      369        3619             90.75
проведение свадьбы             2313      183        2130             92.09

**Вывод:** 

Зависимость между целью кредита и возвратом кредита в срок есть:

- Кредиты на **"операции с недвижимостью"** имеют самый высокий процент возврата кредита в срок - 92.74%.
- Следом за ними хорошо возвращают кредит на **"проведение свадьбы"** (92.09%).
- Самые низкие показатели у категорий клиентов, берущих кредиты на **"получение образования"** (90.75%) и **"операции с автомобилем"** (90.65%).

Объяснить данную ситуацию можно следующим образом:

Клиенты, которые берут **кредит на недвижимость**, обычно очень ответственно относятся к выплатам, так как штрафы за просрочку платежа сильно усугубляют ситуацию. Разница в показателе возврата кредита в срок между операциями с недвижимостью и **проведением свадьбы** не так велик - возможно, кредит на свадьбу не столь большой по сравнению с кредитом на машину, недвижимость и образование, поэтому по нему легче вовремя вносить платежи. Кроме того, никто не отменял финансовую поддержку молодых на свадьбе, что позволяет покрывать часть кредита, уменьшая тем самым его тело. 

Что касается категории **"получение образования"**, то здесь вероятнее всего за учебу платит сам студент, не всегда получается найти высокооплачиваемую работу в этот период времени. 

А с **кредитом на машину** появляется не только ежемесячный платеж, но еще и большой ряд расходов, которые нужно учитывать: покупка шин, страховка, ТО, а случае с б/у автомобилем возникает еще больше трат на расходники.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Опираясь на полученную теорию и датафрейм, пропуски были выявлены в ["Задании 4"](#gap): в столбцах days_employed и total_income. Данные в этих столбцах - количественные, причины их появления могут быть случайными.
1. Например, клиент просто не указал данные о своей зарплате и количестве проработанных дней.
2. Возможны случаи, что предоставление данных необязательно, поэтому появляются пропуски (хотя в этой ситуации кредит невозможно выдать без информации о заработной плате).
3. При возникновении технических проблем, которые появляются на этапе загрузки и выгрузки данных, а также в случае неправильного запроса, объединения таблиц, возможного повреждения файла при сохранении и т.д. 



#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Потому что если использовать метод оценки выборки средним арифметическим, а в данном задании мы убирали пропуски в столбце дохода клиента - количественной переменной, мы столкнемся с некорректными данными, потому что у одного клиента может быть очень низкая зарплата (в нашем примере 20.667), в то время как максимальная зарплата будет - 2.265.604. Существует вариант - избавиться от выбросов, путем усечения данных по 5-10% самых больших и самых маленьких зарплат, а далее работать с новой выборкой, беря среднее арифметическое. Но лучше "справиться" с такими выбросами поможет медиана, которая позволит упорядочить все зарплаты в выборке, определить середину этого ряда. Вероятность того, что там находится самая типичная зарплата - велика.

In [38]:
data['total_income'].min()

20667

In [39]:
data['total_income'].max()

2265604

### Шаг 4: общий вывод.

Проведенная работа доказала существование зависимостей между возвратом кредита в срок и рядом факторов: количество детей, семейное положение, уровень дохода, цели кредита.

В ходе выполнения работы по исследованию надежности заемщиков были поставлены и выполнены следующие задачи:
- определены и удалены пропуски в данных,
- обработаны артефакты (аномалии),
- выявлены дубликаты и приведены в соответствующий требованиям вид,
- доказаны гипотезы путем категоризации данных и составления сводных таблиц для их последующего анализа.

Рекомендации по выбору клиента для последующей выдачи кредита:
- отсутствие детей коррелирует с высоким показателем возврата кредита в срок,
- вдовец или вдова с большей вероятностью будут выплачивать кредит вовремя,
- заработок клиента должен быть не менее 30.000,
- заемщики, берущие кредит на недвижимость, наиболее ответственно подходят к его возврату.